packages/firestore/src/local/local_documents_view.ts

```diff
@@ -431,3 +375,94 @@ export class LocalDocumentsView {
       : PersistencePromise.resolve(MutableDocument.newInvalidDocument(key));
   }
 }
+
+function recalculateAndSaveOverlays(
+  transaction: PersistenceTransaction,
+  docs: MutableDocumentMap,
+  documentOverlayCache: DocumentOverlayCache,
+  serializer: LocalSerializer,
+  userId: string
+): PersistencePromise<void> {
+  const masks = newDocumentKeyMap<FieldMask | null>();
+  // A reverse lookup map from batch id to the documents within that batch.
+  let documentsByBatchId = new SortedMap<number, DocumentKeySet>(
+    (key1: number, key2: number) => key1 - key2
+  );
+  let processed = documentKeySet();
+  return getAllMutationBatchesAffectingDocumentKeys(
+    transaction,
+    docs,
+    serializer,
+    userId
+  )
+    .next(batches => {
+      for (const batch of batches) {
+        batch.keys().forEach(key => {
+          const baseDoc = docs.get(key);
+          if (baseDoc === null) {
+            return;
+          }
+          let mask: FieldMask | null = masks.get(key) || FieldMask.empty();
+          mask = batch.applyToLocalView(baseDoc, mask);
+          masks.set(key, mask);
+          const newSet = (
+            documentsByBatchId.get(batch.batchId) || documentKeySet()
+          ).add(key);
+          documentsByBatchId = documentsByBatchId.insert(batch.batchId, newSet);
+        });
+      }
+    })
+    .next(() => {
+      const promises: Array<PersistencePromise<void>> = [];
+      // Iterate in descending order of batch IDs, and skip documents that are
+      // already saved.
+      const iter = documentsByBatchId.getReverseIterator();
+      while (iter.hasNext()) {
+        const entry = iter.getNext();
+        const batchId = entry.key;
+        const keys = entry.value;
+        const overlays = newMutationMap();
+        keys.forEach(key => {
+          if (!processed.has(key)) {
+            const overlayMutation = calculateOverlayMutation(
+              docs.get(key)!,
+              masks.get(key)!
+            );
+            if (overlayMutation !== null) {
+              overlays.set(key, overlayMutation);
+            }
+            processed = processed.add(key);
+          }
+        });
+        promises.push(
+          documentOverlayCache.saveOverlays(transaction, batchId, overlays)
+        );
+      }
+      return PersistencePromise.waitFor(promises);
+    });
+}
+
+/**
+ * Recalculates overlays by reading the documents from remote document cache
+ * first, and saves them after they are calculated.
+ */
+export function recalculateAndSaveOverlaysForDocumentKeys(
+  transaction: PersistenceTransaction,
+  documentKeys: DocumentKeySet,
+  remoteDocumentCache: RemoteDocumentCache,
+  documentOverlayCache: DocumentOverlayCache,
+  serializer: LocalSerializer,
+  userId: string
```

Comment:
FWIW it is worth, I think we should consider re-ordering the arguments in the methods. I find it easier to read if we first pass the state that is shared (all caches) and then pass the call specific state. To me, this is a better order:

  transaction: PersistenceTransaction,
  remoteDocumentCache: RemoteDocumentCache,
  documentOverlayCache: DocumentOverlayCache,
  serializer: LocalSerializer,
  userId: string
  documentKeys: DocumentKeySet,

Obviously, this is just a gut feeling which can be ignored (especially since I can't actually explain the reasoning well)